pkg/odo/cli/component/common_push.go

```diff
@@ -5,122 +5,12 @@ import (
-	configFlag       bool
-	sourceFlag       bool
-
-	EnvSpecificInfo *envinfo.EnvSpecificInfo
-}
-
-// NewCommonPushOptions instantiates a commonPushOptions object
-func NewCommonPushOptions() *CommonPushOptions {
-	return &CommonPushOptions{}
-}
-
-func (o *PushOptions) SetClientset(clientset *clientset.Clientset) {
-	o.clientset = clientset
-}
-
-//InitEnvInfoFromContext initializes envinfo from the context
-func (cpo *CommonPushOptions) InitEnvInfoFromContext() (err error) {
-	cpo.EnvSpecificInfo, err = envinfo.NewEnvSpecificInfo(cpo.componentContext)
-	if err != nil {
-		return err
-	}
-	return nil
-}
-
-//AddContextFlag adds the context flag to specified command storing value of flag in options.componentContext
-func (cpo *CommonPushOptions) AddContextFlag(cmd *cobra.Command) {
-	odoutil.AddContextFlag(cmd, &cpo.componentContext)
-}
-
-// ResolveSrcAndConfigFlags sets all pushes if none is asked
-func (cpo *CommonPushOptions) ResolveSrcAndConfigFlags() {
-	// If neither config nor source flag is passed, update both config and source to the component
-	if !cpo.configFlag && !cpo.sourceFlag {
-		cpo.configFlag = true
-		cpo.sourceFlag = true
-	}
-}
-
-// ResolveProject completes the push options as needed
-func (cpo *CommonPushOptions) ResolveProject(prjName string) (err error) {
-
-	// check if project exist
-	isPrjExists, err := cpo.clientset.ProjectClient.Exists(prjName)
-	if err != nil {
-		return fmt.Errorf("failed to check if project with name %s exists: %w", prjName, err)
-	}
-	if !isPrjExists {
-		err = cpo.clientset.ProjectClient.Create(prjName, true)
-		if err != nil {
-			return fmt.Errorf(
-				"project %s does not exist. Failed creating it. Please try after creating project using `odo project create <project_name>`: %w",
-				prjName, err,
-			)
-		}
-	}
-	cpo.Context.KClient.SetNamespace(prjName)
-	return
-}
-
-// retrieveKubernetesDefaultNamespace tries to retrieve the current active namespace
-// to set as a default namespace
-func retrieveKubernetesDefaultNamespace() (string, error) {
-	// Get current active namespace
-	client, err := kclient.New()
-	if err != nil {
-		return "", err
-	}
-	return client.Namespace, nil
-}
-
-// retrieveCmdNamespace retrieves the namespace from project flag, if unset
-// we revert to the default namespace available from Kubernetes
-func retrieveCmdNamespace(cmdline cmdline.Cmdline) (componentNamespace string, err error) {
-	// For "odo create" check to see if --project has been passed.
-	if cmdline.IsFlagSet("project") {
-		componentNamespace, err = cmdline.FlagValue("project")
-		if err != nil {
-			return "", err
-		}
-	} else {
-		componentNamespace, err = retrieveKubernetesDefaultNamespace()
-		if err != nil {
-			return "", err
-		}
-	}
-
-	return componentNamespace, nil
-}
-
 // GatherName parses the Devfile and retrieves an appropriate name in two ways.
 // 1. If metadata.name exists, we use it
 // 2. If metadata.name does NOT exist, we use the folder name where the devfile.yaml is located
```

Comment:
Thinking out loud:

This file now has only one function `GatherName`. Does this function belong more in `libdevfile` package? Secondly, and more importantly, I wonder if any devfiles in the default registry don't have `metadata.name`. This function was useful in the past when all devfiles didn't have this field. But now, IMO, it's safe to replace calls to this function with just `devfileObj.Data.GetMetadata().Name`.

I'm not suggesting to address this in this PR, but we could open a separate `good-first-issue` if what I am saying makes sense.

Reply:
I moved it to the `pkg/component` package in the meantime.

I prefer not to move it to libdevfile, as it contains some odo logic IMO.

Reply:
Indeed, indeed. The first part of just getting name is straightforward, doesn't need a separate function; but the second part is odo specific. Although what I was thinking is maybe we could get rid of it and see because devfiles these days do have `.metadaata.name` defined. And this could be a separate issue.